akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkSpec.scala

```diff
@@ -133,4 +140,136 @@ class SinkSpec extends AkkaSpec {
     }
   }
 
+  "Java collector Sink" must {
+    import scala.compat.java8.FunctionConverters._
+
+    class TestCollector(_supplier: () ⇒ Supplier[Array[Int]],
+                        _accumulator: () ⇒ BiConsumer[Array[Int], Int],
+                        _combiner: () ⇒ BinaryOperator[Array[Int]],
+                        _finisher: () ⇒ function.Function[Array[Int], Int]) extends Collector[Int, Array[Int], Int] {
+      override def supplier(): Supplier[Array[Int]] = _supplier()
+      override def combiner(): BinaryOperator[Array[Int]] = _combiner()
+      override def finisher(): function.Function[Array[Int], Int] = _finisher()
+      override def accumulator(): BiConsumer[Array[Int], Int] = _accumulator()
+      override def characteristics(): util.Set[Characteristics] = util.Collections.emptySet()
+    }
+
+    val intIdentity: ToIntFunction[Int] = new ToIntFunction[Int] {
+      override def applyAsInt(value: Int): Int = value
+    }
+
+    def supplier(): Supplier[Array[Int]] = new Supplier[Array[Int]] {
+      override def get(): Array[Int] = Array.ofDim(1)
+    }
+    def accumulator(): BiConsumer[Array[Int], Int] = new BiConsumer[Array[Int], Int] {
+      override def accept(a: Array[Int], b: Int): Unit = a(0) = intIdentity.applyAsInt(b)
+    }
+
+    def combiner(): BinaryOperator[Array[Int]] = new BinaryOperator[Array[Int]] {
+      override def apply(a: Array[Int], b: Array[Int]): Array[Int] = { a(0) += b(0); a }
+    }
+    def finisher(): function.Function[Array[Int], Int] = new function.Function[Array[Int], Int] {
+      override def apply(a: Array[Int]): Int = a(0)
+    }
+
+    "work in the happy case" in {
+      Source(1 to 100).map(_.toString).runWith(Sink.javaCollector(Collectors.joining(", ")))
+        .futureValue should ===((1 to 100).mkString(", "))
+    }
+
+    "work parallelly in the happy case" in {
+      Source(1 to 100).runWith(Sink.javaCollectorParallelUnordered(4)(Collectors.summingInt[Int](intIdentity)))
+        .futureValue should ===(5050)
+    }
+
+    "be reusable" in {
+      val sink = Sink.javaCollector(Collectors.summingInt[Int](intIdentity))
+      Source(1 to 4).runWith(sink).futureValue should ===(10)
+
+      // Collector has state so it preserves all previous elements that went though
+      Source(4 to 6).runWith(sink).futureValue should ===(25)
+    }
+
+    "fail if getting the supplier fails" in {
+      def failedSupplier(): Supplier[Array[Int]] = throw TE("")
+      val future = Source(1 to 100).runWith(Sink.javaCollector(
+        new TestCollector(failedSupplier, accumulator, combiner, finisher)))
+      a[TE] shouldBe thrownBy { Await.result(future, 300.millis) }
```

Comment:
This is too short for some of our CI environments. There is no reason to use a short timeout here. Use 3.seconds when in doubt, or better `remaining` or `remainingOrDefault`. Several more occurrences...